### ISSUE TITLE
Add support for to-many path expressions

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#13639]
jOOQ 3.11 introduced support for "path expressions" that generate implicit joins: https://github.com/jOOQ/jOOQ/issues/1502

The feature was inspired by Hibernate's HQL, which also calls these *"implicit joins"*, but the main feature here is a "path expression" of the form:

```sql
SELECT 
  customer.first_name, 
  customer.last_name, 

  -- Path expression:
  customer.address.city.country.name
FROM customer
```

So far, path expressions are supported only for to-one relationships, because the semantics is straightforward. Path expressions for to-many relationships were postponed (see e.g. https://github.com/jOOQ/jOOQ/issues/12037, https://github.com/jOOQ/jOOQ/issues/14614).

This task aims for supporting them also for to-many relationships, including:

- one-to-many
- many-to-many

And for completeness' sake, also for

- one-to-one (inverse path, see https://github.com/jOOQ/jOOQ/issues/8667)
- many-to-one (inverse path over relationship tables, see https://github.com/jOOQ/jOOQ/issues/13074)

### Specification

The updated specification for what path segments do is this:

- A path segment can go
  - `to-one`: From `child.parent.column`, e.g. `ADDRESS.city().NAME` (existing functionality since 3.11)
  - `one-to-many`: From `parent.child.column`, e.g. `CUSTOMER.payment().AMOUNT` (new functionality, one-to-many)
  - `many-to-many`: From `child1.child2.column`, e.g. `ACTOR.film().TITLE` (new functionality, many-to-many)
- Path segments are named
  - `to-one`: After the parent table by default, foreign key if ambiguous
  - `one-to-many`: TBD!
  - `many-to-many`: TBD!

### Tasks

- [x] API
  - [x] `org.jooq.InverseForeignKey` and related API
  - [x] `org.jooq.meta.InverseForeignKeyDefinition` and related API
- [x] Code generation support
  - [x] Implementation
    - [x] Java
    - [x] Kotlin
    - [x] Scala
  - [x] Add new generator strategy specification for inverse foreign key navigation naming for `one-to-many`
    - [x] The default behaviour is the same as for `to-one` paths: Target table name by default, foreign key name if ambiguous paths
  - [x] Add new generator strategy specification for `many-to-many` navigation naming over 2 FKs pointing to the same parent
    - [x] The default behaviour is to use the target table name by default, outbound foreign key name if ambiguous paths.
  - [x] https://github.com/jOOQ/jOOQ/issues/14993
  - [x] Apply `implicitJoinPathsUseTableNameForUnambiguousFKs` flag as well
  - [x] Ambiguity resolution
    - [x] The simplest solution here is to:
      - [x] Continue generating all `to-one` paths
      - [x] Generate all non-conflicting (with the above) `to-many` paths, and log otherwise
      - [x] Generate all non-conflicting (with the above) `many-to-many` paths, and log otherwise
- [x] Runtime support
  - [x] Define rules in what scope path expressions (explicit or implicit) generate their joins
    - [x] `to-one` path expressions always generate joins at the location of the root declaration
    - [x] `to-many` path expressions must generate joins in a local scope, if the root declaration is in an outer query
      - [x] Test this with several levels of scoping / correlation
  - [x] Add support for using path expressions in `FROM` (https://github.com/jOOQ/jOOQ/issues/14985)
    - For `to-one` path expressions (retrofitting the existing implementation)
      - This can be useful to force a `LEFT JOIN`, `INNER JOIN`, `STRAIGHT_JOIN`, `CROSS JOIN` or whatever
    - For `to-many` path expressions
    - [x] Implicit to-many joins should use the same emulation as in DML statements, using an inline subquery, to prevent unexpected cartesian products from standalone path expressions in non-`FROM` clauses
    - [x] As some users may be fine with implicit to-many joins, we can make the join type configurable: `Settings.renderImplicitJoinToManyType`. Possible values are those of `RenderImplicitJoinType`: `DEFAULT`, `INNER_JOIN`, `LEFT_JOIN`, `SCALAR_SUBQUERY`
    - [x] The existing `Settings.renderImplicitJoinType` flag will also support `SCALAR_SUBQUERY`: https://github.com/jOOQ/jOOQ/issues/15754
    - [x] Both of these flags are effective only if the user doesn't override the join type with an explicit path join.

### Edge cases to test

- [x] Self referencing table, implicit joins in both directions (keys become ambiguous if not direction sensitive)
- [x] `INNER JOIN` vs `LEFT JOIN` semantics of implicit joins (`to-many` are always left joined)
  - [x] Paths must retain local operator precedence, see https://github.com/jOOQ/jOOQ/issues/11010

### Caveats

- [x] DML implicitations
  - [x] The SQL being generated is the same as for to-one path expressions: `(subquery) = value`, which will fail at runtime if the to-many path expression produces more than one value. This seems acceptable
  - [x] For MySQL, the `UPDATE .. JOIN` implementation is skipped in order to produce the same behaviour as for other dialects.

Related work:

- https://github.com/jOOQ/jOOQ/issues/7506
- https://github.com/jOOQ/jOOQ/issues/8667
- https://github.com/jOOQ/jOOQ/issues/12037
- https://github.com/jOOQ/jOOQ/issues/13074
- https://github.com/jOOQ/jOOQ/issues/14985
- https://github.com/jOOQ/jOOQ/issues/14992
- https://github.com/jOOQ/jOOQ/issues/14993

### Documentation
 
In a related discussion, it has become clear that this feature will require extensive documentation in tutorials, etc: https://github.com/jOOQ/jOOQ/issues/13069#issuecomment-1146553040

The problem is that a lot of folks seem to have difficulty understanding the idea behind `LATERAL` or `APPLY`, which are a SQL standard by now, where joins are ordered by their lexical order and thus may be correlated, just like `Stream::flatMap`. `to-many` path expressions are no different. In a way, these two are the same, conceptually:

```sql
-- Using path navigation
SELECT * FROM book, book.authors

-- Using LATERAL
SELECT * FROM book, LATERAL (authors(book.author_id))
```

So, we need:

- [ ] A blog post explaining the concept along the above lines
- [x] Manual sections explaining
  - [x] Path expressions
  - [ ] Various implementations, including
    - [x] Implicit joins for `to-one` path expressions
    - [ ] Correlated subqueries for `to-one` path expressions
    - [x] Explicit joins via path navigations (https://github.com/jOOQ/jOOQ/issues/14985)

### Historic context

A "path expression" as the above can be implemented in at least 2 ways:

**Using implicit joins**

This is the most efficient implementation, as it can re-use `JOIN` trees for shared path segments, in order to join each table only once per path segment.

```sql
SELECT
  customer.first_name, 
  customer.last_name, 
  country.name
FROM customer
JOIN address USING (address_id)
JOIN city USING (city_id)
JOIN country USING (country_id)
```

**Using correlated subqueries**

This is what we did for DML support of path expressions: https://github.com/jOOQ/jOOQ/issues/7508. A correlated subquery is just syntax sugar for a `LEFT JOIN`

```sql
SELECT
  customer.first_name, 
  customer.last_name, 
  (
    SELECT country.name
    FROM country
    JOIN city USING (country_id)
    JOIN address USING (city_id)

    -- Correlation here
    WHERE address.address_id = customer.address_id
  )
FROM customer
```

Or even, using these most canonical solutions of nesting several correlated subqueries, one per path segment, the nesting is in the order of path segments:

```sql
SELECT
  customer.first_name, 
  customer.last_name, 
  (
    SELECT (
      SELECT (
        SELECT country.name
        FROM country
        WHERE country.country_id = city.country_id
      )
      FROM city
      WHERE city.city_id = address.city_id
    )
    FROM address

    -- Correlation here
    WHERE address.address_id = customer.address_id
  )
FROM customer
```

Or also, the nesting is in the inverse order of path segments (this is currently being implemented for DML, because the recursive algorithm was easier this way):

```sql
SELECT
  customer.first_name, 
  customer.last_name, 
  (
    SELECT country.name
    FROM country
    WHERE country.country_id = (
      SELECT city.country_id 
      FROM city
      WHERE city.city_id = (
        SELECT address.address_id
        FROM address
        
        -- Correlation here
        WHERE address.customer_id = customer.customer_id
      )
    )
  )
FROM customer
```

**Comment by alf** [comment]
A couple of thoughts regarding names for the path expression.

Our data model was designed using ORM/NIAM back in the day. This formalism is new to me, but I find that the concept of roles is very useful:

> ORM pictures the world simply in terms of objects (entities or values) that play roles (parts in relationships). For example, you are now playing the role of reading, and this paper is playing the role of being read.
— http://www.orm.net/pdf/jdm99.pdf

The important lesson is that while relationships are symmetrical, the roles the different actors are playing in the relationships are different. I believe that this is just what we need for the names of the path expressions here. We must add some extra configuration on the foreign keys of course, but the result could be great 🤞

**Comment by lukaseder** [comment]
Thanks for your message, @alf. I cannot access that PDF, unfortunately. In any case, yes the labels on the relationships becomes even more important. So far, only the foreign key can have a name, and it always corresponds to what the child calls its parent. We'll need some way to name also the inverse. This can't be derived automatically from metadata (except for the reasonable defaults of just using either the foreign key again, or the child table name). So, users will need to provide this information synthetically in code generation configuration.

**Comment by lukaseder** [comment]
I've been prototyping this over the past days. There are no technical hurdles as far as the code generator is concerned. A simple new concept called "inverse foreign key" will be added with two types:

- `org.jooq.InverseForeignKey` for the runtime model
- `org.jooq.meta.InverseForeignKeyDefinition` for the code generation model

Every `ForeignKey` has an `InverseForeignKey`, which just reverses the arrow, and can be used to decide whether we're traversing:

- a `to-one` path (`ForeignKey`)
- a `to-many` path (`InverseForeignKey`)

An additional type modelling `many-to-many` paths might be considered, later.

Before `to-many` path expressions can be supported, we need to support path expressions in the `FROM` clause. I have a feeling that this is a prerequisite. It's a useful feature for `to-one` path expressions as well, occassionally, when users want to influence the join type explicitly

- https://github.com/jOOQ/jOOQ/issues/14985

**Comment by lukaseder** [comment]
It's really amazing how well #14985 works, almost out of the box! Path segment declarations can be skipped, and will thus produce implicit joins again, if not all paths are declared explicitly. E.g. a query like this:

```java
ctx.select(
        CUSTOMER.ID,
        CUSTOMER.address().city().country().NAME)
   .from(CUSTOMER, CUSTOMER.address().city().country())
```

Will produce a query like this:

```sql
SELECT
  customer.id
  country.name
FROM
  customer
    -- These joins are implicit
    LEFT JOIN address ON customer.address_id = address.address_id
    LEFT JOIN city ON address.city_id = city.city_id,

  -- This cross join is explicit, as per the user query
  country 

-- This join predicate is derived from the explicit path in FROM
WHERE city.country_id = country.country_id
```

**Comment by lukaseder** [comment]
Regarding the possibility of offering this feature as an implicit path join, we'll add a `Settings` attribute, which allows for opting in to implicit to-many path joins. The default being requiring explicit path declarations in `FROM` to avoid confusion about the semantics or accidents.

But from an implementation perspective, the logic is the same as that of `to-one` path joins, which already work out of the box, so the exception is just something to add to the status quo.

**Comment by lukaseder** [comment]
At the same time, there should be a `Settings` that disallows implicit `to-one` path joins, the default being allowing them.

**Comment by lukaseder** [comment]
There are some interesting edge cases of implicit joins when looking at correlated subqueries. E.g. in this cas, the resulting query is obvious:

```java
Author a = AUTHOR.as("a");
ctx.select(
        a.ID,
        field(
            select(countDistinct(a.book().bookToBookStore().bookStore().NAME))
            .from(a.book())
        ))
   .from(a)
```

The correlation `a.book()` happens in the subquery, and thus any path starting from `a.book()` must be joined within the subquery:

```sql
select
  a.ID,
  (
    select count(distinct alias_21445969.NAME)
    from (
      book as alias_33045723
        join (
          book_to_book_store as alias_96285256
            join book_store as alias_21445969
              on alias_96285256.BOOK_STORE_NAME = alias_21445969.NAME
        )
          on alias_96285256.BOOK_ID = alias_33045723.ID
      )
    where alias_33045723.AUTHOR_ID = a.ID
  )
from author as a
```

But users might be inclined to pass the entire path to the `FROM` clause, like this:

```java
Author a = AUTHOR.as("a");
ctx.select(
        a.ID,
        field(
            select(countDistinct(a.book().bookToBookStore().bookStore().NAME))
            .from(a.book().bookToBookStore().bookStore())
        ))
   .from(a)
```

But a canonical implementation won't find all the path segments of the path in the subquery, so it will fall back to joining things to the outer query:

```sql
select
  a.ID,
  (
    select count(distinct alias_21445969.NAME)
    from book_store as alias_21445969
    where alias_96285256.BOOK_STORE_NAME = alias_21445969.NAME
  )
from (
  author as a
    join (
      book as alias_33045723
        join book_to_book_store as alias_96285256
          on alias_96285256.BOOK_ID = alias_33045723.ID
    )
      on alias_33045723.AUTHOR_ID = a.ID
  )
```

This is quite a different query, and I think it's the logical one. The table being declared in the subquery is the `BOOK_STORE` table, not its preceding path segments. `a` is declared in the outer query, and thus all intermediary path segments need to be declared there as well.

Yet another version of the query would do everything implicitly:

```java
Author a = AUTHOR.as("a");
ctx.select(
        a.ID,
        field(
            select(countDistinct(a.book().bookToBookStore().bookStore().NAME))
        ))
   .from(a)
```

This would now produce a meaningless correlated subquery:

```sql
select
  a.ID,
  (
    select count(distinct alias_21445969.NAME)
  )
from (
  t_author as a
    join (
      t_book as alias_33045723
        join (
          t_book_to_book_store as alias_96285256
            join t_book_store as alias_21445969
              on alias_96285256.BOOK_STORE_NAME = alias_21445969.NAME
        )
          on alias_96285256.BOOK_ID = alias_33045723.ID
    )
      on alias_33045723.AUTHOR_ID = a.ID
  )
order by a.ID
```

All joins are added outside of the subquery.

An alternative implementation would add all joins locally until a table is found in scope (i.e. `a`), but that's not what we're currently doing with `to-one` joins. So, I suspect that changing these things would be very backwards incompatible with versions up to jOOQ 3.11.

I'll think about this a bit more as I write more tests.

**Comment by lukaseder** [comment]
The `many-to-many` path expressions can be implemented purely as a code generation feature. E.g. in the Sakila database (https://www.jooq.org/sakila), the `ACTOR <-> FILM` relationship can be navigated as:

```java
ACTOR.film();
```

Which is just syntax sugar for

```java
ACTOR.filmActor().film();
```

In fact, making the two equivalent is desirable so we can re-use join trees that are created implicitly, if any.

**Comment by lukaseder** [comment]
Hmm, maybe the `many-to-many` case isn't so simple. If users use explicit path joins, then the join type should apply for the entire path, not just the last segment.

Well, this yak will need tons of shaving also next week.

**Comment by lukaseder** [comment]
I'm still very much torn between allowing `to-many` implicit path joins by default. There are a few very nice use-cases, e.g.:

```java
TLanguage l = T_LANGUAGE.as("l");

ctx.select(l.ID)
   .from(l)
   // Effectively an anti-join
   .where(l.tBook().ID.isNull())
   .fetch();
```

This is nicer than:

```java
TLanguage l = T_LANGUAGE.as("l");

ctx.select(l.ID)
   .from(l)
   .leftJoin(l.tBook())
   .where(l.tBook().ID.isNull())
   .fetch();
```

**Comment by alf** [comment]
I find the latter example nicer to be honest. It’s much more obvious to me what’s going on.

With that said, I would probably have written this as:

```sql
TLanguage l = T_LANGUAGE.as("l");

ctx.select(l.ID)
   .from(l)
   .leftJoin(l.tBook()).on(l.tBook().ID.isNull())
   .fetch();
```

Reason being that I have problems remembering how sql acts when putting predicates on the right relation in a left join. I believe this can turn the statement into an inner join or something.

**Comment by lukaseder** [comment]
> I find the latter example nicer to be honest. It’s much more obvious to me what’s going on.

I guess you're right 😅. I'm definitely getting carried away here.

> With that said, I would probably have written this as:

That's not the same query. And of course, there's a blog post about this:
https://blog.jooq.org/the-difference-between-sqls-join-on-clause-and-the-where-clause/

**Comment by lukaseder** [comment]
Anyway, the main concern I have is that if you have 2 diverging to-many paths, you'll get a cross product, which may be *very* weird. E.g. this produces a wrong result:

```java
ctx.select(FILM.ID, count(FILM.actor().ID), count(FILM.category().ID))
   .from(FILM)
   .groupBy(FILM.ID)
```

The reason isn't obvious (a cross product between `ACTOR` and `CATEGORY`). It *looks* as though the paths are local to the `count()` functions, but that's not true. They both produce independent left joins. The mistake would have been more clear like this:


```java
ctx.select(FILM.ID, count(FILM.actor().ID), count(FILM.category().ID))
   .from(FILM)
   .leftJoin(FILM.actor())
   .leftJoin(FILM.category())
   .groupBy(FILM.ID)
```

**Comment by alf** [comment]
Thanks for the pointer to your blog about this. As I said, for some reason this is hard for me to grasp. 😅

**Comment by lukaseder** [comment]
Well, the real world use cases of putting non-join predicates in the `ON` clause aren't too many :)

**Comment by lukaseder** [comment]
While writing an integration test, I ran into the scoping problem again from this comment:
https://github.com/jOOQ/jOOQ/issues/13639#issuecomment-1527487983

I ran into it by accident, trying to write a `many-to-many` join (see also https://github.com/jOOQ/jOOQ/issues/8012#issuecomment-1562608457):

```java
TBook b2 = T_BOOK.where(exists(
    selectFrom(T_BOOK.tBookStore())
    .where(T_BOOK.tBookStore().NAME.eq("Ex Libris"))
));

List<Integer> r2 =
create().select(b2.ID)
        .from(b2)
        .orderBy(b2.ID)
        .fetch(b2.ID);
```

The `T_BOOK.tBookStore()` is just an abbreviation for `T_BOOK.tBookToBookStore().tBookStore()`, i.e. a `many-to-many` path join. But this isn't immediately obvious from the method call. Hence, this query was produced:

```sql
select T_BOOK.ID

from (

  T_BOOK
    -- Unintended JOIN here:

    left outer join T_BOOK_TO_BOOK_STORE alias_111750500

      on alias_111750500.BOOK_ID = T_BOOK.ID

  )

where exists (

  select alias_43908512.NAME

  from T_BOOK_STORE alias_43908512

  where (

    alias_43908512.NAME = 'Ex Libris'

    and alias_111750500.BOOK_STORE_NAME = alias_43908512.NAME

  )

)

order by T_BOOK.ID
```

Obviously not what I intended. I wanted this kind of query to be produced, instead:

```sql
select T_BOOK.ID
from T_BOOK
where exists (
  select alias_43908512.NAME
  from T_BOOK_STORE alias_43908512
    left outer join T_BOOK_TO_BOOK_STORE alias_111750500
      on alias_111750500.BOOK_STORE_NAME = alias_43908512.NAME
  where (
    alias_43908512.NAME = 'Ex Libris'
    and alias_111750500.BOOK_ID = T_BOOK.ID
  )
)
order by T_BOOK.ID
```

My intuition tells me that:

- An implicit `to-one` JOIN should always be generated where the parent path is generated
  - This allows for re-using path joins instead of re-joining them every time
  - This was already done this way since jOOQ 3.11, so we won't introduce regressions by keeping things this way
  - `to-one` JOIN never impact the resulting number of rows, because they're `LEFT JOIN` of `0-1` parent rows.
- An implicit `to-many` JOIN should be generated in the local scope when the parent path is declared in an outer query
  - Semantics is more important than re-using path joins, as semantics differ depending on where the joins are generated
  - This is a new feature so no regressions are introduced by designing this one way or another
  - `to-many` JOIN tend to impact the resulting number of rows. A correlated subquery shouldn't impact the outer query

Irrespective of whether implicit `to-many` path joins are enabled by default or not, the feature must be designed consistently in case they are enabled. I'll reflect on this a bit more and will try to implement this improvement today.

**Comment by lukaseder** [comment]
> I'm still very much torn between allowing `to-many` implicit path joins by default. There are a few very nice use-cases, e.g.:

Did another very nice query like this:

```java
ctx.select(A.X, count(A.b().Y))
   .from(A) // Implicit .leftJoin(A.b())
   .groupBy(A.X)
   .orderBy(A.X)
   .fetch();
```

Where `A.b()` is a to-many path. With appropriate naming on the path, it becomes obvious what this does. I'm really still very torn. Allowing implicit to-many paths is:

- Simpler for me to implement 😅 (the checks aren't trivial)
- Very elegant in simple queries

But again, the problem is mostly complex queries, where an accidental cartesian product needs to be debugged. That can be very frustrating if we allow implicit to-many paths by default (power users can definitely turn this on by default. I love it too much myself, to not offer this option).

**Comment by lukaseder** [comment]
I've made up my mind about one of the remaining issues with the `to-many` path expressions, which has been bothering me for a long time. By default, they won't work if they're producing a cartesian product, simply because this would be *very* unexpected. But instead of letting jOOQ throw an exception, we'll delegate the check to SQL by using the existing DML emulation for implicit join paths using scalar subqueries, see:

- #15754
- #15755

We already have a flag `Settings.renderImplicitJoinType`, which allows for enforcing `INNER JOIN` always, even if a foreign key is nullable, effectively:

- Producing a (surprising) filter irrespective of the appearance of the implicit join path (e.g. in `SELECT`, which normally doesn't filter results)
- Emulating Hibernate's behaviour (see https://lists.jboss.org/pipermail/hibernate-dev/2018-February/017299.html)

So, this existing flag is already altering behaviour of implicit to-one joins. We can do the same with a new flag `Settings.renderImplicitJoinToManyType` flag for to-many joins.

Note that either flag is ineffective, if the path join is declared explicitly, in case of which no implicit join path needs to be added.

**Comment by lukaseder** [comment]
So, I'll update the 3.19 manual to reflect the latest changes, and then we're done!

**Comment by lukaseder** [comment]
The default implementation of implicit to-many joins with correlated subqueries helped discover an interesting bug in CockroachDB:

- https://github.com/cockroachdb/cockroach/issues/114227

It shows a risk of the approach, which depends on the sophistication of the optimiser. It's easy to see how CockroachDB can come up with the plan causing a prematurely raised error for a to-many path that produces only one child, but could have produced multiple children with different bind variables.

**Comment by lukaseder** [comment]
Of course, there are dialects that just truncate excess rows from scalar subqueries, e.g. SQLite:
https://twitter.com/lukaseder/status/1724339362823803004

```sql
select (select 1 union select 2)
```

Produces:

```
|(select 1 union select 2)|
|-------------------------|
|1                        |
```

There's no way for jOOQ to detect whether our generated to-many scalar subqueries did produce such excess rows in SQLite. So, perhaps it is better, after all, for jOOQ to throw an exception by default when encountering an implicit to-many join?
